### PR TITLE
feat(parser,codegen): parse and print leading comments

### PR DESCRIPTION
**Pull request by Boshen**
Got a prototype working ... I am going to break this apart into stacked PRs, with proper testing.

NOTE: only special leading comments such as jsdoc, `/* @... */` are printed.

### CODE REVIEW

File: tasks/coverage/codegen_sourcemap.snap
Comment (Boshen) [inline tasks/coverage/codegen_sourcemap.snap:1]: This snapshot is kind of useless.

File: tasks/transform_conformance/babel.snap.md [hunk 1]
```diff
@@ -2489,23 +2465,7 @@ after transform: SymbolId(2): [ReferenceId(4), ReferenceId(6)]
...
-Reference symbol mismatch:
-after transform: ReferenceId(3): Some("AliasModule")
-rebuilt        : ReferenceId(2): Some("AliasModule")
-Reference symbol mismatch:
-after transform: ReferenceId(4): Some("AliasModule")
-rebuilt        : ReferenceId(3): Some("AliasModule")
-Unresolved reference IDs mismatch for "LongNameModule":
-after transform: [ReferenceId(1), ReferenceId(5)]
-rebuilt        : [ReferenceId(1)]
+x Output mismatch
```
Comment (Boshen) [inline tasks/transform_conformance/babel.snap.md:2468]: We need to move the spans ... cc @Dunqing

File: crates/oxc_codegen/src/lib.rs [hunk 2]
```diff
@@ -590,4 +607,27 @@ impl<'a> Codegen<'a> {
     fn try_take_moved_comment(&mut self, node_start: u32) -> Option<Vec<AnnotationComment>> {
         self.move_comment_map.remove(&node_start)
     }
+
+    fn print_leading_comments(&mut self, start: u32) {
+        if self.options.minify {
+            return;
+        }
+        let Some(source_text) = self.source_text else { return };
+        let Some(comments) = self.leading_comments.remove(&start) else { return };
```
Comment (Boshen) [inline crates/oxc_codegen/src/lib.rs:616]: Performance killer :-/ we need to somehow carry over whether a token has leading comments to the ast node, for a fast path.

File: crates/oxc_codegen/src/lib.rs [hunk 3]
```diff
@@ -185,6 +187,20 @@ impl<'a> Codegen<'a> {
...
     ) -> Self {
+        let mut leading_comments: FxHashMap<u32, Vec<Comment>> = FxHashMap::default();
+        for comment in trivias
+            .comments()
+            .copied()
+            .filter(|comment| comment.attached_to != 0 && comment.is_leading())
+            .filter(|comment| {
+                let s = comment.span.source_text(source_text);
+                // only print jsdoc and `@__PURE__`, `@license`, `@preserve` etc
+                (comment.is_multi_line() && s.starts_with('*')) || s.trim_start().starts_with('@')
```
Comment (Boshen) [inline crates/oxc_codegen/src/lib.rs:198]: Need to find all the special comments :-/

File: crates/oxc_ast/src/trivia.rs [hunk 4]
```diff
@@ -8,19 +8,39 @@ use std::{
...
     pub kind: CommentKind,
     /// The span of the comment text (without leading/trailing delimiters).
     pub span: Span,
+
+    pub position: CommentPosition,
+    pub preceded_by_newline: Option<bool>,
+    pub followed_by_newline: Option<bool>,
+
+    pub attached_to: u32,
 }
```
Comment (Boshen) [inline crates/oxc_ast/src/trivia.rs:30]: Need to study what information is required by prettier, cc @leaysgur 

I don't remember where this came from https://github.com/oxc-project/oxc/blob/91bb06e6ff519115f86fab1695a738ab58b875d0/crates/oxc_prettier/src/comments/mod.rs#L54-L65

File: crates/oxc_ast/src/trivia.rs [hunk 5]
```diff
@@ -8,19 +8,39 @@ use std::{
 
 use oxc_span::Span;
 
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum CommentPosition {
+    Leading,
```
Comment (Boshen) [inline crates/oxc_ast/src/trivia.rs:13]: Leading comments are all comments before a token.

File: crates/oxc_ast/src/trivia.rs [hunk 6]
```diff
@@ -8,19 +8,39 @@ use std::{
...
+    Trailing,
```
Comment (Boshen) [inline crates/oxc_ast/src/trivia.rs:14]: Trailing comments are all comments tailing a token and before the newline.

File: crates/oxc_ast/src/trivia.rs (hunk 4 above)
Comment (leaysgur) [reply crates/oxc_ast/src/trivia.rs:30]: I'm still on the way to read through these area, but as far as I can tell are:

- Prettier performs 3 phases to collect comment information
  - 1) Find `preceding`, `enclosing` and|or `following` node for each comment
    - Relatively intuitive, but there is exception
  - 2) Classify comments into `ownLine`, `endOfLine` and `remaining` group
  - 3) Attach comments to each node with 1 + 2
    - This one using so many utils...
- Final AST nodes contain `comments` field which are `ESTree.Comment[]` and `Comment` is extended with `leading` and `trailing` flag.
  - I found a few more extra properties but not understand yet

😅

Comment (leaysgur) [reply crates/oxc_ast/src/trivia.rs:30]: > I don't remember where this came from 

Maybe this?

https://github.com/prettier/prettier/blob/aa3853b7765645b3f3d8a76e41cf6d70b93c01fd/src/language-js/utils/index.js#L1005

**Pull request closed without merging**